README.md
# vzdump
OpenVZ backup scripts

Patches based on latest debian sid/buster: https://packages.debian.org/buster/vzdump  

```
dpkg-source: info: extracting vzdump in vzdump-1.2.6
dpkg-source: info: unpacking vzdump_1.2.6.orig.tar.gz
dpkg-source: info: applying vzdump_1.2.6-5.diff.gz
dpkg-source: info: upstream files that have been modified:
 vzdump-1.2.6/Makefile
 vzdump-1.2.6/OpenVZ.pm
 vzdump-1.2.6/VZDump.pminitial commit vzdump_1.2.6  …
```

## --noarch
did not create a tar archive but keeps the resulting
dump into a directory
this allows real incremental backup

## --mail-error-only
send notification mail to address(es) defined by --mailto only if a dump failed.


ChangeLog
2018-01-06  Simon Stier <github@sist-it.de>

	* vzdump: add option --noarch.
	--noarch did not create a tar archive but keeps the resulting
  dump into a directory
  this allows real incremental backups

2010-05-20  Proxmox Support Team  <support@proxmox.com>

	* vzrestore (restore_openvz): new --force option to overwrite
	existing data (patch from gilles)

2009-10-22  Proxmox Support Team  <support@proxmox.com>

	* VZDump.pm (run_command): correctly detect rsync command (avoid
	backup to fail when files vanished before they could be
	transferred )

2009-10-19  Proxmox Support Team  <support@proxmox.com>

	* vzrestore (restore_openvz): correctly handle '$VEID' macro in
	config files (VE_ROOT, VE_PRIVATE).

	* OpenVZ.pm (read_global_vz_config): do not substitute $VEID

2009-10-13  Proxmox Support Team  <support@proxmox.com>

	* OpenVZ.pm (read_vz_list): use -x (don't cross filesystem
	boundaries) for rsync
	(archive): also use --one-file-system for tar

	* VZDump.pm (run_hook_script): support for new --script option

	* hook-script.pl: example hook script 

2009-10-12  Proxmox Support Team  <support@proxmox.com>

	* VZDump.pm (exec_backup_task): also remove old log files

2009-10-09  Proxmox Support Team  <support@proxmox.com>

	* VZDump.pm: encode backup time in backup filenames

2009-09-21  Proxmox Support Team  <support@proxmox.com>

	* VZDump.pm (storage_info): new --storage option for PVE 

2009-09-08  Proxmox Support Team  <support@proxmox.com>

	* Makefile (vmtar, sparsecp, qmrestore): moved the QemuServer
	related code to the qemu-server package.

2009-09-07  Proxmox Support Team  <support@proxmox.com>

	* Makefile (PACKAGE): increase version to 1.2-1

	* sparsecp.c: small utility to copy stdin to a file preserving
	holes (sparse copy).

2009-09-03  Proxmox Support Team  <support@proxmox.com>

	* vzrestore: split out restore code into new binary 

	* qmrestore: split out restore code into new binary 

2009-09-02  Proxmox Support Team  <support@proxmox.com>

	* OpenVZ.pm (prepare): delay loading config file to this stage.

	* OpenVZ.pm (lock_vm): set lock during backup

	* vzdump.spec.in (Requires): add perl-LockFile-Simple
		
	* control (Depends): add  liblockfile-simple-perl

	* vmtar.c: tar like utility to backup devices.

	* OpenVZ.pm (prepare): use same snapshot name as with
	QemuServer.pm ("vzsnap-$hostname-0").

	* VZDump.pm: we now try to avoid using /var/tmp as temporary storage
	whenever possible. Only use /var/tmp when dumpdir is no posix
	filesystem (or when user explicit specifies --tmpdir).

2009-08-31  Proxmox Support Team  <support@proxmox.com>

	* vzdump.spec.in (Requires): replace MTA with smtpdaemon

2009-08-25  Proxmox Support Team  <support@proxmox.com>

	* rewrite: new plugin architecture. This way we can support plain
	OpenVZ system, but also use QemuServer and other libraries on PVE.

	* Plugin.pm: the base class for plugins.

	* VZDump.pm: contains common code and control logic.

	* OpenVZ: The OpenVZ plugin.
	
	* QemuServer: The PVE QemuServer plugin.
	
	* VZDump.pm (read_vzdump_defaults): use snapshot mode by default
	(it is no longer possible to make online backups without
	stop/suspend/snapshot)

2009-08-03  Proxmox Support Team  <support@proxmox.com>

	* vzdump: set LANG=C to avoid locale warnings

2009-05-29  Proxmox Support Team  <support@proxmox.com>

	* vzdump (get_lvm_device): also return filesystem type
	(snapshot_vm): always mount with '-t $fstype', pass "-o nouuid"
	for xfs filesystems.
	(get_lvm_mapping): use ':' as separator 

2008-09-17  Proxmox Support Team  <support@proxmox.com>

	* vzdump (snapshot_vm): use --no-whole-file --inplace for rsync (works
	better for large files), also use --stats --numeric-id and
	apply bandwidth limits with --bwlimit 
	(archive_vm): use --sparse option for tar

	* vzdump: support to backup/restore qemu-server images

	* vzdump: new --stdexcludes option (we no longer exclude files by
	default, because most usres expect to have a full backup including
	all files and logs)

	* vzdump: add --tmpdir option as suggested by Ugo Bellavance
	
	* vzdump: add --wait option, using locbal lock file to
	make sure only one instance is running
	
	* vzdump: set default snapshot size to 1GB, new --size option to
	specify snapshot size

	* copyright: fixed fsf address

	* Makefile (SBINDIR): set to /usr/sbin (command is only useful for 
	user root)

	* vzdump.spec.in, control.in: we now depend on rsync, cstream and MTA

2008-09-16  Proxmox Support Team  <support@proxmox.com>

	* vzdump (read_vzdump_defaults): added global configuration file.

2008-08-18  Proxmox Support Team  <support@proxmox.com>

	* vzdump: added --node option (proxmox ve cluster support),
	use --rsyncable for compressed files
	removed --xdelta option 

2007-06-28  Proxmox Support Team  <support@proxmox.com>

	* vzdump: use values of vz.conf if no private/root dir specified
	in XXX.conf

2007-05-25  Proxmox Support Team  <support@proxmox.com>

	* vzdump (run_command): better parser
	(read_global_config): improved parser
	(send_mail): allow multiple receivers (multiple --mailto options)
	(send_mail): add additional status infos

2007-05-24  Proxmox Support Team  <support@proxmox.com>

	* vzdump (send_mail): generate nicer mails

2007-05-21  Proxmox Support Team  <support@proxmox.com>

	* vzdump: new --exclude-path option, use 'find' to avoid tar
	sockets warnings, try to skip all log files by default, use tar
	option --ignore-failed-read

2007-05-10  Proxmox Support Team  <support@proxmox.com>

	* vzdump: additionally log time, new --exclude option (suggested
	by Ugo Bellavance)

2007-04-04  Proxmox Support Team  <support@proxmox.com>

	* vzdump (get_device): use -P with df (fixes bug reported by
	Jernej Proenta)

2007-03-26  Proxmox Support Team  <support@proxmox.com>

	* vzdump (read_global_config): read config of VEs >= 1000
	- issue warning when doing backup without suspend/snapshot

2007-03-22  Proxmox Support Team  <support@proxmox.com>


Plugin.pm
package PVE::VZDump::Plugin;

#    Copyright (C) 2007-2009 Proxmox Server Solutions GmbH
#
#    Copyright: vzdump is under GNU GPL, the GNU General Public License.
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; version 2 dated June, 1991.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the
#    Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
#    MA 02110-1301, USA.
#
#    Author: Dietmar Maurer <dietmar@proxmox.com>

use strict;
use warnings;

sub set_logfd {
    my ($self, $logfd) = @_;

    $self->{logfd} = $logfd;
}

sub cmd {
    my ($self, $cmdstr, $input) = @_;

    return PVE::VZDump::run_command ($self->{logfd}, $cmdstr, $input);   
}

sub cmd_noerr {
    my ($self, $cmdstr, $input) = @_;

    my $res;
    eval { $res = $self->cmd ($cmdstr, $input); };
    $self->logerr ($@) if $@;
    return $res;
}

sub loginfo {
    my ($self, $msg) = @_;

    PVE::VZDump::debugmsg ('info', $msg, $self->{logfd}, 0);
}

sub logerr {
    my ($self, $msg) = @_;

    PVE::VZDump::debugmsg ('err', $msg, $self->{logfd}, 0);
}

sub type {
    return 'unknown';
};

sub vmlist {
    my ($self) = @_;

    return [ keys %{$self->{vmlist}} ] if $self->{vmlist};

    return [];
}

sub vm_status {
    my ($self, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub prepare {
    my ($self, $task, $vmid, $mode) = @_;

    die "internal error"; # implement in subclass
}

sub lock_vm {
    my ($self, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub unlock_vm {
    my ($self, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub stop_vm {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub start_vm {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub suspend_vm {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub resume_vm {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub snapshot {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub copy_data_phase2 {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub assemble {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

sub archive {
    my ($self, $task, $vmid, $filename) = @_;

    die "internal error"; # implement in subclass
}

sub cleanup {
    my ($self, $task, $vmid) = @_;

    die "internal error"; # implement in subclass
}

1;


hook-script.pl
#!/usr/bin/perl -w

# example hook script for vzdump (--script option)

use strict;

print "HOOK: " . join (' ', @ARGV) . "\n";

my $phase = shift;

if ($phase eq 'job-start' || 
    $phase eq 'job-end'  || 
    $phase eq 'job-abort') { 

    # do what you want 

} elsif ($phase eq 'backup-start' || 
	 $phase eq 'backup-end' ||
	 $phase eq 'backup-abort' || 
	 $phase eq 'pre-stop' ||
	 $phase eq 'pre-restart') {

    print "HOOK: " . join (' ', @ARGV) . "\n";

    my $mode = shift; # stop/suspend/snapshot

    my $vmid = shift;

    my $vmtype = $ENV{VMTYPE}; # openvz/qemu

    my $dumpdir = $ENV{DUMPDIR};

    my $hostname = $ENV{HOSTNAME};

    my $tarfile = $ENV{TARFILE};

    my $logfile = $ENV{LOGFILE};

    print "HOOK-ENV: vmtype=$vmtype;dumpdir=$dumpdir;hostname=$hostname;tarfile=$tarfile;logfile=$logfile\n";

    # example: copy resulting files to another host using scp
    #if ($phase eq 'backup-end') {
    #	system ("scp $tarfile $logfile backup-host:/backup-dir") == 0 ||
    #	    die "copy to backup-host failed";
    #	unlink $tarfile;
    #	unlink $logfile;
    #}
    
} else {

    die "got unknown phase '$phase'";

}

exit (0);


OpenVZ.pm
package PVE::VZDump::OpenVZ;

#    Copyright (C) 2007-2009 Proxmox Server Solutions GmbH
#
#    Copyright: vzdump is under GNU GPL, the GNU General Public License.
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; version 2 dated June, 1991.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the
#    Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
#    MA 02110-1301, USA.
#
#    Author: Dietmar Maurer <dietmar@proxmox.com>

use strict;
use warnings;
use File::Path;
use File::Basename;
use PVE::VZDump;
use Sys::Hostname;
use LockFile::Simple;

use base qw (PVE::VZDump::Plugin);

use constant SCRIPT_EXT => qw (start stop mount umount);
use constant VZDIR => '/etc/vz';

my $remove_quotes = sub {
    my $str = shift;

    $str =~ s/^\s*\"?//;
    $str =~ s/\"?\s*$//;

    return $str;
};

# read global vz.conf
sub read_global_vz_config {

    local $/;

    my $res = {
	rootdir => '/var/lib/vz/root/$VEID', # note '$VEID' is a place holder
	privatedir => '/var/lib/vz/private/$VEID', # note '$VEID' is a place holder
	dumpdir => '/var/lib/vz/dump',
	lockdir => '/var/lib/vz/lock',
    };
    
    my $filename = VZDIR . "/vz.conf";

    my $fh = IO::File->new ($filename, "r");
    return $res if !$fh;
    my $data = <$fh> || '';
    $fh->close();

    if ($data =~ m/^\s*VE_PRIVATE=(.*)$/m) {
	my $dir = &$remove_quotes ($1);
	if ($dir !~ m/\$VEID/) {
	    warn "VE_PRIVATE does not contain '\$VEID' ('$dir')\n";
	} else {
	    $res->{privatedir} = $dir;
	}
    }
    if ($data =~ m/^\s*VE_ROOT=(.*)$/m) {
	my $dir = &$remove_quotes ($1);
	if ($dir !~ m/\$VEID/) {
	    warn "VE_ROOT does not contain '\$VEID' ('$dir')\n";
	} else {
	    $res->{rootdir} = $dir;
	}
    }
    if ($data =~ m/^\s*DUMPDIR=(.*)$/m) {
	my $dir = &$remove_quotes ($1);
	$dir =~ s|/\$VEID$||;
	$res->{dumpdir} = $dir;
    }
    if ($data =~ m/^\s*LOCKDIR=(.*)$/m) {
	my $dir = &$remove_quotes ($1);
	$res->{lockdir} = $dir;
    }

    return $res;
}

my $load_vz_conf = sub {
    my ($self, $vmid) = @_;

    local $/;

    my $conf = $self->{vmlist}->{$vmid}->{conffile};

    my $fh = IO::File->new ($conf, "r") ||
	die "unable to open config file '$conf'\n";
    my $data = <$fh>;
    $fh->close();

    my $dir;
    if ($data =~ m/^\s*VE_PRIVATE=(.*)$/m) {
	$dir = &$remove_quotes ($1);
    } else {
	$dir = $self->{privatedir};
    }
    $dir =~ s/\$VEID/$vmid/;
    $self->{vmlist}->{$vmid}->{dir} = $dir;

    if ($data =~ m/^\s*HOSTNAME=(.*)/m) {
	$self->{vmlist}->{$vmid}->{hostname} = &$remove_quotes ($1);
    } else {
	$self->{vmlist}->{$vmid}->{hostname} = "VM $vmid";
    }
};

sub read_vz_list {

    my $vmlist = {};

    my $dir = VZDIR . "/conf";
    foreach my $conf (<$dir/*.conf>) {

	next if $conf !~ m|/(\d\d\d+)\.conf$|;

	my $vmid = $1;

	$vmlist->{$vmid}->{conffile} = $conf;
    }

    return $vmlist;
}

my $rsync_vm = sub {
    my ($self, $task, $from, $to, $text) = @_;

    $self->loginfo ("starting $text sync $from to $to");

    my $starttime = time();

    my $opts = $self->{vzdump}->{opts};

    my $rsyncopts = "--stats -x --numeric-ids --bwlimit=$opts->{bwlimit}";

    $self->cmd ("rsync $rsyncopts -aH --delete --no-whole-file --inplace '$from' '$to'");

    my $delay = time () - $starttime;

    $self->loginfo ("$text sync finished ($delay seconds)");
};

sub new {
    my ($class, $vzdump) = @_;
    
    PVE::VZDump::check_bin ('vzctl');

    my $self = bless read_global_vz_config ();

    $self->{vzdump} = $vzdump;

    $self->{vmlist} = read_vz_list ();

    return $self;
};

sub type {
    return 'openvz';
}

sub vm_status {
    my ($self, $vmid) = @_;

    my $status_text = $self->cmd ("vzctl status $vmid");
    chomp $status_text;

    my $running = $status_text =~ m/running/ ? 1 : 0;
   
    return wantarray ? ($running, $status_text) : $running; 
}

sub prepare {
    my ($self, $task, $vmid, $mode) = @_;

    $self->$load_vz_conf ($vmid);

    my $dir = $self->{vmlist}->{$vmid}->{dir};

    my $diskinfo = { dir => $dir };

    $task->{hostname} = $self->{vmlist}->{$vmid}->{hostname};

    $task->{diskinfo} = $diskinfo;

    my $hostname = hostname(); 

    if ($mode eq 'snapshot') {

	my $lvmmap = PVE::VZDump::get_lvm_mapping();
	my ($srcdev, $lvmpath, $lvmvg, $lvmlv, $fstype) =
	    PVE::VZDump::get_lvm_device ($dir, $lvmmap);

	my $targetdev = PVE::VZDump::get_lvm_device ($task->{dumpdir}, $lvmmap);

	die ("mode failure - unable to detect lvm volume group\n") if !$lvmvg;
	die ("mode failure - wrong lvm mount point '$lvmpath'\n") if $dir !~ m|/?$lvmpath/?|;
	die ("mode failure - unable to dump into snapshot (use option --dumpdir)\n") 
	    if $targetdev eq $srcdev;

	$diskinfo->{snapname} = "vzsnap-$hostname-0";
	$diskinfo->{snapdev} = "/dev/$lvmvg/$diskinfo->{snapname}";
	$diskinfo->{srcdev}  = $srcdev;
	$diskinfo->{lvmvg}   = $lvmvg;
	$diskinfo->{lvmlv}   = $lvmlv;
	$diskinfo->{fstype}  = $fstype;
	$diskinfo->{lvmpath} = $lvmpath;
	$diskinfo->{mountpoint} = "/mnt/vzsnap0";
	    
	$task->{snapdir} = $dir;
	$task->{snapdir} =~ s|/?$lvmpath/?|$diskinfo->{mountpoint}/|;
    
    } elsif ($mode eq 'suspend') {
	$task->{snapdir} = $task->{syncdir};
    } else {
	$task->{snapdir} = $dir;
    }
}

sub lock_vm {
    my ($self, $vmid) = @_;

    my $filename = "$self->{lockdir}/$vmid.lck";

    my $lockmgr = LockFile::Simple->make(-format => '%f',
					 -autoclean => 1,
					 -max => 30, 
					 -delay => 2, 
					 -stale => 1,
					 -nfs => 0);

    $self->{lock} = $lockmgr->lock($filename) || die "can't lock VM $vmid\n";
}

sub unlock_vm {
    my ($self, $vmid) = @_;

    $self->{lock}->release();
}

sub copy_data_phase1 {
    my ($self, $task) = @_;

    $self->$rsync_vm ($task, "$task->{diskinfo}->{dir}/", $task->{snapdir}, "first");
}

sub stop_vm {
    my ($self, $task, $vmid) = @_;

    $self->cmd ("vzctl --skiplock stop $vmid");
}

sub start_vm {
    my ($self, $task, $vmid) = @_;

    $self->cmd ("vzctl --skiplock start $vmid");
}

sub suspend_vm {
    my ($self, $task, $vmid) = @_;

    $self->cmd ("vzctl --skiplock chkpnt $vmid --suspend");
}

sub snapshot {
    my ($self, $task) = @_;

    my $opts = $self->{vzdump}->{opts};

    my $di = $task->{diskinfo};

    mkpath $di->{mountpoint}; # create mount point for lvm snapshot

    if (-b $di->{snapdev}) {
	$self->loginfo ("trying to remove stale snapshot '$di->{snapdev}'");
	    
	$self->cmd_noerr ("umount $di->{mountpoint}");
	    
	$self->cmd_noerr ("lvremove -f $di->{snapdev}");
    }

    $self->loginfo ("creating lvm snapshot of $di->{srcdev} ('$di->{snapdev}')");

    $task->{cleanup}->{lvm_snapshot} = 1;
	
    $self->cmd ("lvcreate --size $opts->{size}M --snapshot" .
		" --name $di->{snapname} /dev/$di->{lvmvg}/$di->{lvmlv}");

    my $mopts = $di->{fstype} eq 'xfs' ? "-o nouuid" : '';

    $task->{cleanup}->{snapshot_mount} = 1;

    $self->cmd ("mount -t $di->{fstype} $mopts $di->{snapdev} $di->{mountpoint}");
}

sub copy_data_phase2 {
    my ($self, $task) = @_;

    $self->$rsync_vm ($task, "$task->{diskinfo}->{dir}/", $task->{snapdir}, "final");
}

sub resume_vm {
    my ($self, $task, $vmid) = @_;

    $self->cmd ("vzctl --skiplock chkpnt $vmid --resume");
}

sub assemble {
    my ($self, $task, $vmid) = @_;

    my $conffile = $self->{vmlist}->{$vmid}->{conffile};

    my $dir = $task->{snapdir};

    if ($self->{vzdump}->{opts}->{noarch}) {
        $task->{cleanup}->{etc_vzdump} = 0;
    } else {
        $task->{cleanup}->{etc_vzdump} = 1;
    }
	
    mkpath "$dir/etc/vzdump/";
    $self->cmd ("cp '$conffile' '$dir/etc/vzdump/vps.conf'");
    my $cfgdir = dirname ($conffile);
    foreach my $s (SCRIPT_EXT) {
	my $fn = "$cfgdir/$vmid.$s";
	$self->cmd ("cp '$fn' '$dir/etc/vzdump/vps.$s'") if -f $fn;
    } 
}

sub archive {
    my ($self, $task, $vmid, $filename) = @_;
    
    my $findexcl = $self->{vzdump}->{findexcl};
    my $findargs = join (' ', @$findexcl) . ' -print0';
    my $opts = $self->{vzdump}->{opts};

    my $srcdir = $self->{vmlist}->{$vmid}->{dir};
    my $snapdir = $task->{snapdir};

    my $bwl = $opts->{bwlimit}*1024; # bandwidth limit for cstream
    my $zflag = $opts->{compress} ? 'z' : '';

    my $taropts = "--totals --sparse --numeric-owner --no-recursion --ignore-failed-read --one-file-system";

    if ($snapdir eq $task->{syncdir} && $snapdir =~ m|^$opts->{dumpdir}/|) {
	$taropts .= " --remove-files"; # try to save space
    }

    my $out = ">$filename";
    $out = "|cstream -t $bwl $out" if $opts->{bwlimit};

    $self->cmd ("(cd $snapdir;find . $findargs|" .
		"tar c${zflag}pf - $taropts --null -T - $out )");

}

sub cleanup {
    my ($self, $task, $vmid) = @_;

    my $di = $task->{diskinfo};

    if ($task->{cleanup}->{snapshot_mount}) {
	$self->cmd_noerr ("umount $di->{mountpoint}");
    }

    if ($task->{cleanup}->{lvm_snapshot}) {
	$self->cmd_noerr ("lvremove -f $di->{snapdev}") if -b $di->{snapdev};
    }

    if ($task->{cleanup}->{etc_vzdump}) {
	my $dir = "$task->{snapdir}/etc/vzdump";
	eval { rmtree $dir if -d $dir; };
	$self->logerr ($@) if $@;
    }
}

1;
